reinstall-cli/src/users.rs

```diff
@@ -0,0 +1,106 @@
+use anyhow::{Context, Result};
+use std::fmt::Display;
+use std::fmt::Formatter;
+use uzers::os::unix::UserExt;
+
+#[derive(Clone, Debug)]
+pub(crate) struct UserKeys {
+    pub(crate) user: String,
+    pub(crate) authorized_keys: String,
+    pub(crate) authorized_keys_path: String,
+}
+
+impl UserKeys {
+    pub(crate) fn num_keys(&self) -> usize {
+        self.authorized_keys.lines().count()
+    }
+}
+
+impl Display for UserKeys {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "User {} ({} authorized keys)",
+            self.user,
+            self.num_keys()
+        )
+    }
+}
+
+/// # Safety
+/// See safety information in [`uzers::all_users`]. We are not impacted because we're not
+/// multi-threaded.
+pub(crate) fn get_all_users_keys() -> Result<Vec<UserKeys>> {
+    #[allow(unsafe_code)]
+    let iter = unsafe { uzers::all_users() };
+
+    let mut result = Vec::new();
+
+    for user_info in iter {
+        if user_info.uid() < 1000 && user_info.uid() != 0 {
+            tracing::debug!(
+                "Skipping user {} with uid {} because it's a system user (uid < 1000)",
+                user_info.name().to_string_lossy(),
+                user_info.uid()
+            );
+            continue;
+        }
+
+        if ["/bin/false", "/bin/true", "nologin"]
+            .into_iter()
+            .any(|x| user_info.shell().to_string_lossy().contains(x))
+        {
+            tracing::debug!(
+                "Skipping user {} with shell {} because it can't login",
+                user_info.name().to_string_lossy(),
+                user_info.shell().to_string_lossy()
+            );
+            continue;
+        }
+
+        let home_dir = user_info.home_dir();
+        let user_authorized_keys_path = home_dir.join(".ssh/authorized_keys");
```

Comment:
It's a bit harder than this in the general case; in CoreOS derivatives we've pushed for https://github.com/coreos/ssh-key-dir for example.

Reply:
But what do we do on the target machine? Try to emulate the same structure as the original machine? Just copy all the keys and dump them naively in the target machine? 

How do we make sure the binary behind `AuthorizedKeysCommand` even behaves the same on the source and target machines?

Since `AuthorizedKeysCommand` is an opaque box, it's actually impossible to know what we would have to do on the target machine such that it still outputs the same keys

This all seems like a bad idea the more I think about it

I think we might have to special case the coreos scenario, as solving this in the general seems literally impossible

Reply:
That seems reasonably straightforward. @germag and I had a live chat about this too and my take is:

- For this initial version, we can aim to tightly constrain our starting points. If we even just handle the stock ssh configuration for e.g. RHEL9.5 and RHEL CoreOS (already enough diversity there since `AuthorizedKeysCommand` is in play in one but not the other) 
- If we even need to add distribution/OS detection and bail out on untested setups (say ubuntu or whatever) I think that's fine to start, with maybe a `--allow-untested-ssh` CLI escape hatch) and obviously ideally it's reasonably easy for us or other folks to add more cases later

Reply:
Long comment only 3 minutes after my second comment, just to make sure - please make sure you've seen my second comment

Reply:
Followup in #1081